### PR TITLE
Enable multicore support for indirect access

### PR DESCRIPTION
**Pull request by shruthip-venkatesh**
#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Makes multi-core gather (`out = x[i]`) correct under work division — for **stick-aligned and partial index counts alike**.

A gather is lowered to an identity op over an index tensor, a shared value table, and an output. The work-division planner ranks dimensions by size and splits the largest. When a value-data dimension (K/N — e.g. an embedding's
hidden dim, a KV cache's head dim) is the largest, the planner split it across cores. Because the value table is shared (every core must be able to gather any row), splitting a value-data dim is incorrect: the output advances its per-core base across the dimension but the shared value tensor does not, so every core reads column 0 of the table and writes it into its own output slice. The result is silently wrong (no crash) and was masked by constant-filled tests.

This PR:
1. **Forbids splitting value-data dims** (the correctness fix) —  `shared_indirect_data_syms`, consulted via `indirect_forbidden_split_syms` and applied through `forbidden_split_syms` in `_default_split`, so the planner divides a gather along the index dimension only.
2. **Pins the shared value tensor to a fixed base** — `SDSCArgs.shared_base`, honored in `core_idx_to_slice_offset`, so the value base does not drift per core.
3. **Adds a value-table span guard** — surfaces the value tensor to the 256 MB per-core span check (`collect_indirect_value_tds`, IndexLoad-aware `get_per_core_span`); conservative/defensive.
4. **Fixes the split round-trip for indirect reads** — `_first_non_indirect_read_index`, so the coeff-keyed split encoding/decoding stays stable when an indirect read is present.
5. **Stick-aligns the index-entry split for partial-last-stick counts** — the index-entry dim is the index tensor's stick dim (32 int32 entries per128-byte stick). Work division splits it in whole sticks, but an even per-core slice of a partial last stick (e.g. 40 = one full stick + 8) straddles the index stick boundary, which the backend cannot step across within a core — so such counts previously miscompiled or fell back to a single core. `enforce_indirect_access_layout._pad_output_for_stick_aligned_split` grows the gather output's entry-dim `device_size` up to the next stick multiple (logical size unchanged; the D2H copy extracts the logical rows from the larger allocation), and `superdsc` grows the SDSC index-entry iteration to match **before** computing per-core strides so the base lands stick-aligned, not mid-stick. `indirect_forbidden_split_syms` enforces the invariant: a partial-stick entry split is forbidden **unless the output was provably padded** to a whole stick (`indirect_entry_output_dim`).

Gathers now parallelize correctly along the index dimension for both stick-aligned and partial index counts. Parallelizing the value-data dims (for small-index workloads like MoE / paged attention) is future work.

#### Which issue(s) this PR is related to:

Issue : https://github.com/torch-spyre/torch-spyre/issues/2725

#### Special notes for your reviewer:
- The load-bearing change is Pt. 1 (forbid value-data-dim splits). Pt. 2 – Pt. 4 are supporting/hardening. Pt. 3 (span guard) is intentionally conservative — a 512 MB value table read correctly in hardware testing, so the per-core span limit's applicability to indirect value tensors isn't firmly established; the guard only logs a critical message, it doesn't fail compilation.
- Pt. 5 pads the gather **output**, not the index — the index's logical count is never changed, which is why `indirect_forbidden_split_syms` reads that (unpadded) count to forbid, and lifts the forbiddance only after seeing the
padded output extent. The add-then-discard is deliberate: forbid by default, split only when the output is provably stick-aligned.
- **Trade-off:** when the index dim is too small to absorb all cores, the gather falls back to fewer cores rather than splitting a value-data dim. A partial count that can't be padded likewise stays single-core. Correct-but-serial is intended; silent corruption is not.
- Tests sweep `SENCORES` including a non-power-of-two count (6), assert the split map against the planner's `core_split(ceil(n/32), cores)`, and add stick-aligned and partial-stick gather scenarios. 

#### Additional note:
Full multi-core **scatter** (`out[i] = src`) is not yet enabled by this PR — the IndexLoad is on the output there, and it needs an index-uniqueness guarantee (future work). The partial-stick guard (Pt. 5) does cover scatter **defensively**: a scatter's in-place destination can't be padded, so a partial-stick scatter correctly stays single-core rather than miscompiling. Parallelizing value-data dims is also future work.

### CODE REVIEW

File: torch_spyre/_inductor/work_division.py [hunk 1]
```diff
@@ -907,7 +1064,13 @@ def work_distribution_pass(
             return
 
     splits, output_dims, reduction_dims = _default_split(
-        it_space_adjusted, output_td, committed_splits, max_cores, symbol_meta
+        it_space_adjusted,
+        output_td,
+        committed_splits,
+        max_cores,
+        symbol_meta,
+        forbidden_split_syms=shared_indirect_data_syms(op),
```
Comment (kaoutar55) [inline torch_spyre/_inductor/work_division.py]: Blocker: this is the only place `forbidden_split_syms` is applied. The mandatory `span_reduction_pass` (line 863) runs earlier and calls `must_split_vars` (line 885) then `apply_splits` without the forbidden dims (`must_split_vars` takes no such argument). For a gather whose output is over the 256 MB per-core span while the index dim is small, span reduction can split a value-data dim to meet the limit, and `work_distribution_pass` then reads that as a committed split (see the docstring at line 998) and does not undo it. That brings back the wrong addressing this PR fixes. Pass `forbidden_split_syms` into `span_reduction_pass` and `must_split_vars` (and fall back to fewer cores), or document why span reduction never splits a shared-table data dim.

File: torch_spyre/_inductor/pass_utils.py [hunk 2]
```diff
@@ -436,22 +436,85 @@ def _build_indirect_load_subs(op: ComputedBuffer) -> dict[sympy.Symbol, sympy.Ex
...
+) -> "dict[sympy.Symbol, sympy.Expr]":
+    """Find all indirect accesses in an operation and mark them appropriately.
+
+    Called before scheduling to identify which dimensions are accessed through
+    runtime indices. Handles both gather (indirect reads) and scatter (indirect
+    writes). Returns substitutions that mark these dimensions with IndirectAccess
+    so the work division planner knows not to split them incorrectly.
+    """
+    raw = _build_indirect_load_subs(op)
+    raw.update(_build_indirect_store_subs(op))
```
Comment (kaoutar55) [inline torch_spyre/_inductor/pass_utils.py]: Blocker: this will not build after rebase. On main, `_build_indirect_load_subs` returns a tuple `(subs, sizes)` (`pass_utils.py:503`), but here `indirect_access_subs_from_op` does `raw = _build_indirect_load_subs(op)` then `raw.update(...)`, which fails on a tuple. gh shows the PR as `CONFLICTING`. Rebase onto main, update to the new indirect-subs API, and re-run CI, since the green checks ran on an old base.

File: docs/source/user_guide/examples/paged_attention_compute.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
+import torch
```
Comment (kaoutar55) [inline docs/source/user_guide/examples/paged_attention_compute.py:15]: `paged_attention_compute.py` and `paged_attention_kernel.py` are missing the 14-line Apache header that every source file needs (both start with `import torch`). `gather_multicore_exp.py` and `scatter_multicore_exp.py` have it. These two files are not referenced by the new doc and look like debug scratch (bare try/except printing tracebacks, emoji in output). Add the header and link them from the docs, or drop them.

File: torch_spyre/_inductor/codegen/superdsc.py [hunk 4]
```diff
@@ -554,6 +555,7 @@ def _create_sdsc_tensors(
                 arg_index=arg.arg_index,
                 is_index_tensor=is_idx_tensor,
                 related_value_tensor_idx=related_val_idx,
+                shared_base=has_indirect_access and is_indirect_value_tensor(arg),
```
Comment (kaoutar55) [inline torch_spyre/_inductor/codegen/superdsc.py]: Question: `shared_base = has_indirect_access and is_indirect_value_tensor(arg)` assumes `is_indirect_value_tensor` returns True for the scatter destination, which is an output arg, not an input value table. Does it? If not, the destination base still advances per core. An assertion in the new test that `shared_base` is set on the destination would confirm it, since scatter end-to-end is xfail.

**Review by kaoutar55 — CHANGES_REQUESTED** [review]
Sound approach, clear docs. Three blockers: the no-split rule is missing from the mandatory `span_reduction_pass`, there is no multicore test, and the branch conflicts with main. Details below.

### Notes not attached to a diff line

- `tests/inductor/test_indirect_access_gather.py`: Blocker: this is a multicore change with no multicore test. `test_indirect_access_gather.py` and `test_indirect_access_scatter.py` both run at `SENCORES=1` and are not changed. The only check is the example scripts under `docs/source/user_guide/examples/`, which need real hardware and pytest does not collect them. Add tests under `tests/inductor/` that run gather and overwrite-scatter at `SENCORES > 1` across a few core counts and both stick-aligned and unaligned data dims, and assert the split is on the index dim, never on K or N.

File: docs/source/user_guide/examples/paged_attention_compute.py (hunk 3 above)
Comment (shruthip-venkatesh) [reply docs/source/user_guide/examples/paged_attention_compute.py:15]: Added the headers and linked these paged attention examples into _indirect_access_work_division.md_ document.

File: torch_spyre/_inductor/work_division.py (hunk 1 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/work_division.py]: `must_split_vars` : added `forbidden_split_syms: set[Symbol] | None = None`. The _split_vars_ list inside the per-coord loop now filters out any symbol in this set, so the exhaustive search never even considers splitting a shared-table data dimension.

`span_reduction_pass` : computes `forbidden = shared_indirect_data_syms(op) or None` and passes it as `forbidden_split_syms=forbidden` into _must_split_vars_. This closes the gap identified: before this, work_distribution_pass had the guard but span_reduction_pass (which runs first) did not.

File: torch_spyre/_inductor/codegen/superdsc.py (hunk 4 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/codegen/superdsc.py]: I have added the required comments for clarity and also added additional test case.

File: torch_spyre/_inductor/pass_utils.py (hunk 2 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/pass_utils.py]: Done

**Comment by shruthip-venkatesh** [comment]
@kaoutar55 ,
I have addressed the review comments, could you take a look again ?

**Comment by shruthip-venkatesh** [comment]
Depends on https://github.com/torch-spyre/torch-spyre/pull/3409

File: torch_spyre/_inductor/codegen/superdsc.py [hunk 5]
```diff
@@ -758,6 +760,14 @@ def _create_sdsc_tensors(
                 is_index_tensor=is_idx_tensor,
                 related_value_tensor_idx=related_val_idx,
                 per_tile_fixed=arg.per_tile_fixed,
+                # shared_base is True for:
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/codegen/superdsc.py]: As discuss a few weeks back, the shared base portion of this PR should not be required. Do you see any issues without this? If so, we can debug!

Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/codegen/superdsc.py]: Yes, removing this do not cause any issue. I have removed it and updated the test case to cover additional scenarios.

File: torch_spyre/_inductor/enforce_indirect_access_layout.py [hunk 6]
```diff
@@ -53,11 +53,58 @@
...
     device_coordinates,
     indirect_info_from_op,
+    indirect_entry_output_dim,
 )
+from . import config
 
 logger = get_inductor_logger("enforce_indirect_access_layout")
 
 
+def _pad_output_for_stick_aligned_split(op: ComputedBuffer) -> bool:
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/enforce_indirect_access_layout.py:63]: Could we split the padding support into a separate PR from the core work division logic?

File: tests/inductor/indirect_access_common.py [hunk 7]
```diff
@@ -257,6 +257,13 @@ def indirect_access_target_names(op_specs) -> set:
 UNIMPLEMENTED = "unimplemented_op"  # Hit an UnimplementedOp
 NO_SPYRE_OP = "no_spyre_op_spec"  # Fell back to CPU
 
+# Core counts to sweep in the multicore variants of the indirect-access suites.
+# 1 keeps the original single-core coverage; the rest exercise the work-division
+# planner (index-dim splitting, shared-table / scatter-row never-split
+# constraints) at every supported SENCORES value. 6 is included as a
+# non-power-of-two count so uneven core-vs-stick divisions are covered too.
+MULTICORE_SENCORES = (1, 2, 4, 6, 8, 16, 32)
```
Comment (avery-blanchard) [inline tests/inductor/indirect_access_common.py:265]: I think this explodes the indirect access test files without adding much additional coverage. For gather, 97 methods × 7 core configurations (1,2,4,6,8,16,32). Could we keep this like the rest of our tests and just run with 32 cores? If there are edge cases to watch out for, could we add targeted test cases for these scenarios?

File: torch_spyre/_monkey_patch.py [hunk 8]
```diff
@@ -40,22 +40,13 @@ def _patch_tensor_for_spyre():
...
-                )
-            if "device=" in s:
-                return s.replace("device='cpu'", f"device='{self.device}'")
-            if s.endswith(")"):
-                s = s[:-1] + f", device='{self.device}')"
-            else:
-                # Odd case: just append device info
-                s = s + f" (device='{self.device}')"
-            return s
+            # Do NOT call self.to("cpu") here — it can trigger a SIGABRT
```
Comment (avery-blanchard) [inline torch_spyre/_monkey_patch.py]: Did you hit errors with this? I'm unsure how this is related to the rest of the PR contents

File: tests/inductor/indirect_access_common.py (hunk 7 above)
Comment (shruthip-venkatesh) [reply tests/inductor/indirect_access_common.py:265]: Addressed by splitting the test cases.

File: torch_spyre/_monkey_patch.py (hunk 8 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_monkey_patch.py]: Removed now. Not facing error any error now.

File: torch_spyre/_inductor/pass_utils.py [hunk 9]
```diff
@@ -934,6 +991,227 @@ def try_device_coordinates(
...
+    ``out_extent`` is a whole multiple of ``eps``.
+    """
+
+    stick_var: sympy.Symbol  # iteration symbol indexing the entry dim
+    eps: int  # index tensor elems_per_stick (the split granularity)
+    out_pos: int  # entry dim's device position in the output layout
+    out_extent: int  # output device_size at out_pos (physical extent)
+
+
+def indirect_entry_output_dim(op) -> "IndirectEntryDim | None":
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/pass_utils.py]: Suggestion: can we eliminate `IndirectEntryDim` and inline the padding detection?
Its currently used in two places, both of which already have access to the op and the layouts. Instead of bundling this into a data structure, we could replace `indirect_entry_output_dim(op)` with helper like `is_output_stick_aligned_for_entry(op) -> bool` that both the padding pass and the work split logic can call. This eliminates unnecessary indirection and keeps the coordinate-finding logic localized to one place.

Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/pass_utils.py]: Eliminated IndirectEntryDim with helper functions.

**Review by avery-blanchard — DISMISSED** [review]
thanks, looks good to me!

File: torch_spyre/_inductor/enforce_indirect_access_layout.py (hunk 6 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/enforce_indirect_access_layout.py:63]: This changes are integral part of this PR.
Details are discussed offline.

**Review by avery-blanchard — APPROVED** [review]

**Review by kaoutar55 — APPROVED** [review]
Approving. All three blockers from my earlier review are resolved. Thanks!
